p2panda-rs/src/atomic/entry.rs

```diff
@@ -0,0 +1,220 @@
+
+    /// Used log for this entry.
+    log_id: LogId,
+
+    /// Message payload of entry, can be deleted.
+    message: Option<Message>,
+
+    /// Sequence number of this entry.
+    seq_num: SeqNum,
+}
+
+/// Error types for methods of `Entry` struct.
+#[allow(missing_copy_implementations)]
+#[derive(Error, Debug)]
+pub enum EntryError {
+    /// Links should not be set when first entry in log.
+    #[error("backlink and skiplink not valid for this sequence number")]
+    InvalidLinks,
+}
+
+impl Entry {
+    /// Validates and returns a new instance of `Entry`.
+    pub fn new(
+        log_id: &LogId,
+        message: &Message,
+        entry_hash_skiplink: Option<&Hash>,
+        entry_hash_backlink: Option<&Hash>,
+        previous_seq_num: Option<&SeqNum>,
+    ) -> Result<Self> {
+        // If it is the first entry set sequence number to 1, otherwise increment
+        let seq_num = match previous_seq_num {
+            None => SeqNum::default(),
+            Some(s) => {
+                let mut next_seq_num = s.clone();
+                next_seq_num.next().unwrap()
+            }
+        };
+
+        let entry = Self {
+            log_id: log_id.clone().to_owned(),
+            message: Some(message.clone().to_owned()),
+            entry_hash_skiplink: entry_hash_skiplink.cloned(),
+            entry_hash_backlink: entry_hash_backlink.cloned(),
+            seq_num,
+        };
+
+        entry.validate()?;
+
+        Ok(entry)
+    }
+
+    /// Signs the Bamboo entry via Ed25519 key pair and returns the hex-encoded representation.
+    pub fn sign_and_encode(&self, key_pair: &KeyPair) -> Result<EntryEncoded> {
+        // Generate message hash
+        // @TODO: Handle case where message is not set
+        let message_encoded = self.message.clone().unwrap().encode().unwrap();
+        let message_hash = message_encoded.hash();
+        let message_size = message_encoded.size();
+
+        // Convert entry links to Bamboo crate types
+        let backlink = self
+            .entry_hash_backlink
+            .clone()
+            .map(|link| YamfHash::Blake2b(link.to_bytes()));
+
+        let lipmaa_link = self
+            .entry_hash_skiplink
+            .clone()
+            .map(|link| YamfHash::Blake2b(link.to_bytes()));
+
+        // Create bamboo entry. See: https://github.com/AljoschaMeyer/bamboo#encoding for encoding
+        // details and definition of entry fields.
+        let mut entry: BambooEntry<_, &[u8]> = BambooEntry {
+            log_id: self.log_id.as_u64(),
+            is_end_of_feed: false,
+            payload_hash: YamfHash::Blake2b(message_hash.to_bytes()),
+            payload_size: message_size,
+            author: PublicKey::from_bytes(&key_pair.public_key_bytes())?,
+            seq_num: self.seq_num.as_u64(),
+            backlink,
+            lipmaa_link,
+            sig: None,
+        };
+
+        // @TODO: Sign BambooEntry and encode it
+        EntryEncoded::new("dummy")
```

Comment:
I just looked into it and want to write down what I saw to talk about it tomorrow:

So here we want to sign the bamboo-core entry that we created just above. We should be able to use `key_pair` for that, which is a parameter for this function.

However, looking at the [dogs](https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/) I see that a `ed25519_dalek::Keypair` instance is required to sign bytes. In our `KeyPair` implementation, the actual `KeyPair` is thrown away in `::new` and there is no way to recover it from the `public` and `private` halfs as far as I can tell.

Do we need to change our `KeyPair` implementation so that instead of serializing the private key it's serializing the whole keypair? Because a serialized keypair can be turned into `KeyPair` instance again, which we need in order to sign data.

Reply:
I think `KeyPair` should be refactored to allow for a test like

```rust
#[test]
    fn signing() {
        let key_pair = KeyPair::new();
        let message = b"test";
        let signature = key_pair.sign(message);
        assert!(key_pair.verify(message, signature).is_ok())
    }
```